### PR TITLE
[FLINK-18934] Idle stream does not advance watermark in connected stream

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

Make two and multi-input operators respect stream idleness.

## Brief change log

* extract a reusable `CombinedWatermark` class
* drop `StreamStatusMaintainer` and `StreamStatusProvider` interfaces
* Let the StreamStatus flow through the input/output chains

Two open questions:
* do we want to maintain the behaviour from the removed tests? Those tests assumed that there can be records flowing when a stream is IDLE. In such cases we dropped any flowing Watermarks
* Do we want to put the logic for making the stream temporary ACTIVE/IDLE in the `RecordWriterOutput` or do we want to find a better place for it.

## Verifying this change

Adjusted tests that were verifying the idleness, e.g.
* MultipleInputStreamTaskTest#testWatermarkAndStreamStatusForwarding
* TwoInputStreamTaskTest#testWatermarkAndStreamStatusForwarding

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java [hunk 1]
```diff
@@ -158,6 +160,32 @@ public AbstractStreamOperatorV2(StreamOperatorParameters<OUT> parameters, int nu
...
+                                try {
+                                    processWatermark(new Watermark(watermark.getTimestamp()));
+                                } catch (Exception e) {
+                                    throw new WrappingRuntimeException(e);
+                                }
+                            }
+
+                            @Override
+                            public void markIdle() {
+                                streamStatusMaintainer.toggleStreamStatus(StreamStatus.IDLE);
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java:180]: This is the part of the whole solution that I am unsure the most. With this approach the bookkeeping happens both on the operator and stream task level. As far as I can tell it would work with this method empty.

Theoretically we could also move the whole bookkeeping logic from the StreamTask level to the operator level.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java [hunk 2]
```diff
@@ -268,7 +296,11 @@ protected boolean isUsingCustomRawKeyedState() {
      * @throws Exception An exception in this method causes the operator to fail.
      */
     @Override
-    public void open() throws Exception {}
+    public void open() throws Exception {
+        this.watermarkOutputMultiplexer =
+                setupWatermarkMultiplexer(
+                        containingTask.getStreamStatusMaintainer(), numberOfInputs);
+    }
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java:303]: Why are you initialising this non final field in the open method? It probably should happen in the constructor and `watermarkOutputMultiplexer` should be final?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java (hunk 1 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java:180]: What logic on the `StreamTask` level are you referring to?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java (hunk 2 above)
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java:303]: Fully agree, the problem is that the `containingTask.getStreamStatusMaintainer()` is not set until the `open` method.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java (hunk 1 above)
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java:180]: Everything that is in the `StreamStatusMaintainer(s)`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java [hunk 3]
```diff
@@ -243,6 +240,13 @@ public void setup(
         stateKeySelector2 = config.getStatePartitioner(1, getUserCodeClassloader());
     }
 
+    private CombinedWatermark setupCombinedWatermark() {
+        CombinedWatermark combinedWatermark = new CombinedWatermark();
+        combinedWatermark.add(new CombinedWatermark.PartialWatermark());
+        combinedWatermark.add(new CombinedWatermark.PartialWatermark());
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java:246]: shouldn't that depend on the number of inputs? Can't we create that in ctor?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java [hunk 4]
```diff
@@ -628,22 +632,42 @@ public void processWatermark(Watermark mark) throws Exception {
...
+
+    private void emitStreamStatus(StreamStatus streamStatus, int index) {
+        boolean wasIdle = combinedWatermark.isIdle();
+        combinedWatermark.get(index).setIdle(streamStatus.isIdle());
+        if (combinedWatermark.updateCombinedWatermark()) {
+            output.emitWatermark(new Watermark(combinedWatermark.getCombinedWatermark()));
         }
+        if (wasIdle != combinedWatermark.isIdle()) {
+            output.emitStreamStatus(streamStatus);
+        }
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java:662]:
```
if (combinedWatermark.updateStatus(index, streamStatus.isIdle())) {
  output.emitStreamStatus(streamStatus);
}
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java [hunk 5]
```diff
@@ -137,7 +132,7 @@ public AbstractStreamOperatorV2(StreamOperatorParameters<OUT> parameters, int nu
             output = countingOutput;
         }
 
-        latencyStats =
+        this.latencyStats =
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java:135]: unnecessary change

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java [hunk 6]
```diff
@@ -146,6 +141,7 @@ public AbstractStreamOperatorV2(StreamOperatorParameters<OUT> parameters, int nu
         executionConfig = parameters.getContainingTask().getExecutionConfig();
         userCodeClassLoader = parameters.getContainingTask().getUserCodeClassLoader();
         cancelables = parameters.getContainingTask().getCancelables();
+        this.combinedWatermark = setupCombinedWatermark(numberOfInputs);
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java:144]:
```
combinedWatermark = new CombinedWatermark(numberOfInputs);
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamSourceContexts.java [hunk 7]
```diff
@@ -149,15 +142,16 @@ public void close() {}
...
         private AutomaticWatermarkContext(
                 final Output<StreamRecord<T>> output,
                 final long watermarkInterval,
                 final ProcessingTimeService timeService,
                 final Object checkpointLock,
-                final StreamStatusMaintainer streamStatusMaintainer,
                 final long idleTimeout) {
 
-            super(timeService, checkpointLock, streamStatusMaintainer, idleTimeout);
+            super(timeService, checkpointLock, idleTimeout);
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamSourceContexts.java:154]: Removing stream status maintainer feels like an independent change. Can we move that into a separate commit? It makes reviewing harder.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/RecordWriterOutput.java [hunk 8]
```diff
@@ -101,32 +97,51 @@ public void collect(StreamRecord<OUT> record) {
     }
 
     private <X> void pushToRecordWriter(StreamRecord<X> record) {
-        serializationDelegate.setInstance(record);
+        if (announcedStatus.isIdle()) {
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/RecordWriterOutput.java:100]: I think we need to add a comment here to explain why this can happen.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/RecordWriterOutput.java [hunk 9]
```diff
@@ -48,20 +47,19 @@
...
 
-    private final StreamStatusProvider streamStatusProvider;
-
     private final boolean supportsUnalignedCheckpoints;
 
     private final OutputTag outputTag;
 
     private final WatermarkGauge watermarkGauge = new WatermarkGauge();
 
+    private StreamStatus announcedStatus = StreamStatus.ACTIVE;
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/RecordWriterOutput.java:56]: Is active the default throughout whole Flink?

**Review by AHeise — CHANGES_REQUESTED** [review]
Thank you very much for your contribution. The code definitively feels much cleaner with the additional abstraction of `CombinedWatermark`.

However, I'd probably not expose the inner class at all and get rid of the List ancestor. You now have a weird data structure that is unnecessarily mutable in many ways.

I'm also not sure if the tests are sufficient: I'm missing tests that explicitly test the behavior of 2- and N-ary operators. Most tests have been traditionally on task level but that is probably not the best place anymore.

File: flink-core/src/main/java/org/apache/flink/api/common/eventtime/CombinedWatermark.java [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        if (minimumOverAllOutputs > combinedWatermark) {
+            combinedWatermark = minimumOverAllOutputs;
+            return true;
+        }
+
+        return false;
+    }
+
+    /** Per-output watermark state. */
+    public static class PartialWatermark {
```
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-core/src/main/java/org/apache/flink/api/common/eventtime/CombinedWatermark.java:114]: I agree. If I were writing the class from scratch I'd aim for immutability. I don't want to change the logic too much in the crucial `WatermarkOutputMultiplexer`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java (hunk 3 above)
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java:246]: I can add an additional ctor. (I was actually thinking about it) However, we still need the `add` method for the `WatermarkOutputMultiplexer`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/RecordWriterOutput.java (hunk 9 above)
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/RecordWriterOutput.java:56]: I'd say yes. We start Flink cluster with all channels active, and we wait for Watermarks from all inputs.

**Comment by dawidwys** [comment]
I tried addressing your comments. I am still not 100% sure about the short ACTIVE/IDLE cycle or should we rather let records be generated but halt watermarks forwarding.

Do you mind taking another look @AHeise ?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java (hunk 3 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java]: Hm, haven't thought about the dynamic cases (actually Kafka is not really dynamic but I can see that other sources are). Still the comment stands that I wouldn't expose `PartialWatermark`. 

You could have a similar method to `register` in `CombinedWatermark` that returns the index of the newly added partial. The index is then used in `private final Map<String, Integer> watermarkIndexPerOutputId;` This index can then be used to update the `CombinedWatermark` via index.

This approach makes more sense if you can easily have all side-effects inside the `CombinedWatermark#updateWatermark(int index, Watermark watermark)`. So it avoids the case, where you update the partial and then tell the combined watermark: "hey something changed, please check".

However, the current way may still be valid, if you explicitly want to have the information drift and only update at specific times for performance reasons. I'm currently not in the position to assess that properly.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/RecordWriterOutput.java (hunk 8 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/RecordWriterOutput.java:103]: The comment is good in respect to explaining the invariant (you could also give a link to the docs). However, it's not helping the reader to understand why this case is happening.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/RecordWriterOutput.java [hunk 11]
```diff
@@ -101,32 +97,57 @@ public void collect(StreamRecord<OUT> record) {
...
+        serializationDelegate.setInstance(record);
         try {
             recordWriter.emit(serializationDelegate);
         } catch (Exception e) {
             throw new RuntimeException(e.getMessage(), e);
         }
+
+        if (announcedStatus.isIdle()) {
+            writeStreamStatus(StreamStatus.IDLE);
+        }
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/RecordWriterOutput.java:116]: Can we factor out some reusable pattern?

```
try (announcedStatus.ensureActive(this::writeStreamStatus)) {
        serializationDelegate.setInstance(record);
        try {
            recordWriter.emit(serializationDelegate);
        } catch (Exception e) {
            throw new RuntimeException(e.getMessage(), e);
        }
}
```

**Comment by AHeise** [comment]
> I tried addressing your comments. I am still not 100% sure about the short ACTIVE/IDLE cycle or should we rather let records be generated but halt watermarks forwarding.
> 
> Do you mind taking another look @AHeise ?

Looks already quite good. As discussed offline, I think that the ACTIVE/IDLE cycle is a good start and we should just optimize some operators to make them cycling in batches (asyncIO). To that end, could you try to factor out some small reusable pattern? I gave an idea on the respective lines.

Other comments:
* I have not yet giving up on making `PartialWatermark` private though ;) Please check and refute my idea.
* I'd squash the squash commit.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java (hunk 3 above)
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java]: So there are two things here. My comment in this thread referred just to the `add` method, not exposing the `PartialWatermark`. Sure, if the issue was just about the `add` method, I would not expose the `PartialWatermark`.

The second problem is the combination of `PartialWatermark` and `updateCombinedWatermark`. Actually, you wrote the exact reason why we need it:
> However, the current way may still be valid if you explicitly want to have the information drift and only update at specific times for performance reasons. I'm currently not in the position to assess that properly.

That's exactly what happens in the  `WatermarkOutputMultiplexer`. It exposes two access patterns:
* getImmediateOutput
* getDeferredOutput

One immediately updates the combined watermark, and the other one does not. AFAIK the deferred one is used for periodic watermarks. Bear in mind that the `PeriodicWatermark` has the default scope. It is exposed only to the `org.apache.flink.api.common.eventtime` package for it to be used by `WatermarkOutputMultiplexer`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/RecordWriterOutput.java (hunk 11 above)
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/RecordWriterOutput.java:116]: Nice one! Will do.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java (hunk 3 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java]: Okay I'd still think that you could also hide that within `#updateWatermark(..., boolean deferred)` but the benefit of encapsulation gets even smaller (if it even still exists). So leave as is.

**Review by AHeise — APPROVED** [review]
LGTM now ;)
